timer/node_timer.go

```diff
@@ -0,0 +1,143 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022 Datadog, Inc.
+
+package timer
+
+import (
+	"errors"
+	"fmt"
+	"time"
+)
+
+// nodeTimer is the type used for all the timers that can (and will) have children
+type nodeTimer struct {
+	baseTimer
+	components
+
+	// parentComponent is the component stored in the map of the parent timer. We keep a reference to it to update its spent time
+	// once the timer is stopped. It may be nil if the timer is not a child of a tree timer, like when calling NewTimer.
+	parentComponent *component
+
+	// parent is the parent timer. It is used to propagate the stop of the timer to the parent timer and get the remaining time in case the budget has to be inherited.
+	parent NodeTimer
+}
+
+var _ NodeTimer = (*nodeTimer)(nil)
+
+// NewTreeTimer creates a new Timer with the given options. You have to specify either the option WithBudget or WithUnlimitedBudget and at least one component using the WithComponents option.
+func NewTreeTimer(options ...Option) (NodeTimer, error) {
+	config := newConfig(options...)
+	if config.Budget == InheritedBudget {
+		return nil, errors.New("root timer cannot inherit parent budget, please provide a budget using timer.WithBudget() or timer.WithUnlimitedBudget()")
+	}
+
+	if len(config.Components) == 0 {
+		return nil, errors.New("NewTreeTimer: tree timer must have at least one component, otherwise use NewTimer()")
+	}
+
+	return &nodeTimer{
+		baseTimer: baseTimer{
+			config:    config,
+			timeCache: newTimeCache(),
+		},
+		components: newComponents(config.Components),
+	}, nil
+}
+
+func (timer *nodeTimer) NewNode(name string, options ...Option) (NodeTimer, error) {
+	config := newConfig(options...)
+	if len(config.Components) == 0 {
+		return nil, errors.New("NewNode: node timer must have at least one component, otherwise use NewLeaf()")
+	}
```

Comment:
This looks like a very artificial limitation, although I guess the return type difference justifies that...